q2_demux/_demux.py

```diff
@@ -0,0 +1,124 @@
+                    (barcode.metadata['id'].rsplit('/', 1)[0],
+                     sequence.metadata['id'].rsplit('/', 1)[0]))
+            if barcode.metadata['description'].rsplit('/', 1)[0] != \
+               sequence.metadata['description'].rsplit('/', 1)[0]:
+                raise ValueError(
+                    'Mismatched sequence descriptions: %s and %s' %
+                    (barcode.metadata['description'].rsplit('/', 1)[0],
+                     sequence.metadata['description'].rsplit('/', 1)[0]))
+            yield barcode, sequence
+
+    def __next__(self):
+        return next(self.barcode_generator), next(self.sequence_generator)
+
+
+def emp(seqs: BarcodeSequenceIterator,
+        barcodes: qiime.MetadataCategory,
+        rev_comp_barcodes: bool=False,
+        rev_comp_mapping_barcodes: bool=False) \
+        -> SingleLanePerSampleSingleEndFastqDirFmt:
+    barcode_map = {}
+    barcode_len = None
+    for sample_id, barcode in barcodes.to_series().iteritems():
+        if barcode_len is None:
+            barcode_len = len(barcode)
+        elif len(barcode) != barcode_len:
+            raise ValueError('Barcodes of different lengths were detected: '
+                             '%d != %d. Variable length barcodes are not '
+                             'supported.' % (len(barcode), barcode_len))
+        if rev_comp_mapping_barcodes:
+            barcode = str(skbio.DNA(barcode).reverse_complement())
+        if barcode in barcode_map:
+            raise ValueError('A duplicate barcode was detected. The barcode '
+                             '%s was observed for samples %s and %s.'
+                             % (barcode, sample_id, barcode_map[barcode]))
+        barcode_map[barcode] = sample_id
+
+    result = SingleLanePerSampleSingleEndFastqDirFmt()
+
+    manifest = FastqManifestFormat()
+    manifest_fh = manifest.open()
+    manifest_fh.write('sample-id,filename,direction\n')
+
+    per_sample_fastqs = {}
+
+    for barcode, sequence in seqs:
+        if rev_comp_barcodes:
+            barcode = barcode.reverse_complement()
+        barcode = str(barcode)
+        barcode = barcode[:barcode_len]
+
+        try:
+            sample_id = barcode_map[barcode]
+        except KeyError:
+            # TODO: this should ultimately be logged, but we don't currently
+            # have support for that.
+            continue
+
+        if sample_id not in per_sample_fastqs:
+            # The barcode id, lane number and read number are not relevant
+            # here. We might ultimately want to use a dir format other than
+            # SingleLanePerSampleSingleEndFastqDirFmt which doesn't care
+            # about this information. Similarly, the direction of the read
+            # isn't relevant here anymore.
+            barcode_id = len(per_sample_fastqs) + 1
+            path = result.sequences.path_maker(sample_id=sample_id,
+                                               barcode_id=barcode_id,
```

Comment:
The existing regex would not match these files in that case. I'd rather leave this as-is for now and just get a better DirectoryFormat in place eventually which doesn't include these values.